taichi/ui/backends/vulkan/gui.cpp

```diff
@@ -0,0 +1,182 @@
+
+    VkSubmitInfo end_info = {};
+    end_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+    end_info.commandBufferCount = 1;
+    end_info.pCommandBuffers = &command_buffer;
+    vkEndCommandBuffer(command_buffer);
+    vkQueueSubmit(app_context->graphics_queue(), 1, &end_info, VK_NULL_HANDLE);
+
+    vkDeviceWaitIdle(app_context->device());
+    vkFreeCommandBuffers(app_context->device(), app_context->command_pool(), 1,
+                         &command_buffer);
+    ImGui_ImplVulkan_DestroyFontUploadObjects();
+  }
+}
+
+void Gui::create_descriptor_pool() {
+  VkDescriptorPoolSize pool_sizes[] = {
+      {VK_DESCRIPTOR_TYPE_SAMPLER, 1000},
+      {VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, 1000},
+      {VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1000},
+      {VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, 1000},
+      {VK_DESCRIPTOR_TYPE_UNIFORM_TEXEL_BUFFER, 1000},
+      {VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER, 1000},
+      {VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 1000},
+      {VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1000},
+      {VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC, 1000},
+      {VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC, 1000},
+      {VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT, 1000}};
+  VkDescriptorPoolCreateInfo pool_info = {};
+  pool_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO;
+  pool_info.flags = VK_DESCRIPTOR_POOL_CREATE_FREE_DESCRIPTOR_SET_BIT;
+  pool_info.maxSets = 1000 * IM_ARRAYSIZE(pool_sizes);
+  pool_info.poolSizeCount = (uint32_t)IM_ARRAYSIZE(pool_sizes);
+  pool_info.pPoolSizes = pool_sizes;
+  VkResult err = vkCreateDescriptorPool(app_context_->device(), &pool_info,
+                                        VK_NULL_HANDLE, &descriptor_pool_);
+}
+
+void Gui::prepare_for_next_frame() {
+  ImGui_ImplVulkan_NewFrame();
+  ImGui_ImplGlfw_NewFrame();
+  ImGui::NewFrame();
+  is_empty = true;
+}
+
+float Gui::abs_x(float x) {
+  return x * app_context_->swap_chain.swap_chain_extent.width;
+}
+float Gui::abs_y(float y) {
+  return y * app_context_->swap_chain.swap_chain_extent.height;
+}
+
+void Gui::begin(std::string name, float x, float y, float width, float height) {
+  ImGui::SetNextWindowPos(ImVec2(abs_x(x), abs_y(y)), ImGuiCond_Once);
+  ImGui::SetNextWindowSize(ImVec2(abs_x(width), abs_y(height)), ImGuiCond_Once);
+  ImGui::Begin(name.c_str());
+  is_empty = false;
+}
+void Gui::end() {
+  ImGui::End();
+}
+void Gui::text(std::string text) {
+  ImGui::Text(text.c_str());
+}
+bool Gui::checkbox(std::string name, bool old_value) {
+  ImGui::Checkbox(name.c_str(), &old_value);
+  return old_value;
+}
+float Gui::slider_float(std::string name,
+                        float old_value,
+                        float minimum,
+                        float maximum) {
+  ImGui::SliderFloat(name.c_str(), &old_value, minimum, maximum);
+  return old_value;
+}
+glm::vec3 Gui::color_edit_3(std::string name, glm::vec3 old_value) {
+  ImGui::ColorEdit3(name.c_str(), (float *)&old_value);
```

Comment:
this old_value is in stack, this might cause memory problems due to how ImGUI works.

Reply:
Could you please elaborate? I'm under the impression that it sufffices to ensure that `&old_value` is valid through out the call to `ImGui::ColorEdit3`